pkg/admission/namespacelifecycle/admission.go

```diff
@@ -0,0 +1,148 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/apiserver/pkg/admission/plugin/namespace/lifecycle"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// PluginName indicates the name of admission plug-in
+	PluginName = "WorkspaceNamespaceLifecycle"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return newLifcycle()
+	})
+}
+
+// workspaceNamespaceLifecycle is a wrapper of kubernetes namespace lifecycle admission control
+// It uses legacy namespace lifecycle admission control by default, and ignore
+// immortal namespaces when the workspaces is deleting. This can ensure we can remove all immortal
+// namespaces when workspaces is deleting.
+type workspaceNamespaceLifecycle struct {
+	// legacyNamespaceLifecycle is the kube legacy namespace lifecycle
+	legacyNamespaceLifecycle *lifecycle.Lifecycle
+
+	// namespaceLifecycle is used only when workspace is deleting
+	namespaceLifecycle *lifecycle.Lifecycle
+
+	*admission.Handler
+	workspaceLister tenancyv1alpha1lister.ClusterWorkspaceLister
+}
+
+func newLifcycle() (*workspaceNamespaceLifecycle, error) {
+	legacyLifecycle, err := lifecycle.NewLifecycle(sets.NewString(metav1.NamespaceDefault, metav1.NamespaceSystem, metav1.NamespacePublic))
+	if err != nil {
+		return nil, err
+	}
+
+	lifecycle, err := lifecycle.NewLifecycle(sets.NewString())
+	if err != nil {
+		return nil, err
+	}
+
+	return &workspaceNamespaceLifecycle{
+		Handler:                  admission.NewHandler(admission.Create, admission.Update, admission.Delete),
+		legacyNamespaceLifecycle: legacyLifecycle,
+		namespaceLifecycle:       lifecycle,
+	}, nil
+}
+
+var _ = kcpinitializers.WantsKcpInformers(&workspaceNamespaceLifecycle{})
+var _ = initializer.WantsExternalKubeInformerFactory(&workspaceNamespaceLifecycle{})
+var _ = initializer.WantsExternalKubeClientSet(&workspaceNamespaceLifecycle{})
+
+// Admit makes an admission decision based on the request attributes
+func (l *workspaceNamespaceLifecycle) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	// prevent deletion of immortal namespaces
+	if a.GetOperation() != admission.Delete || a.GetKind().GroupKind() != corev1.SchemeGroupVersion.WithKind("Namespace").GroupKind() {
+		return l.legacyNamespaceLifecycle.Admit(ctx, a, o)
+	}
+
+	clusterName, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	org, hasParent := clusterName.Parent()
+	if !hasParent {
+		return l.legacyNamespaceLifecycle.Admit(ctx, a, o)
+	}
+
+	workspaceKey := clusters.ToClusterAwareKey(org, clusterName.Base())
+
+	workspace, err := l.workspaceLister.Get(workspaceKey)
```

Comment:
this won't scale. I would rather prefer some way to decide locally, e.g. by looking at the request user, to decide whether we want to allow deletion.

Background: as soon as we get sharding you won't have a `workspaceLister` anymore that holds all workspaces, only those on the same shard.

Reply:
Or let's say this admission plugin must not fail normal namespace deletion when the shard with the workspace object is down. During workspace deletion though, failing in that case is ok.

Reply:
good point! updated.